nmea_parser.c

```diff
@@ -1,191 +1,315 @@
 #include "nmea_parser.h"
 #include "def.h"
 #include "utils.h"
+#include <sys/types.h>
+#include <sys/socket.h>
 
 extern int g_debug;
 extern int g_foreground;
 extern FILE *fp_console;
+static unsigned char status=0x26;
 
-void parse_NMEA_sensor(char* message, t_sensor_context* sensors)
+// Return a 0 if checksum is bad
+// Return the location of the '*' character if checksum is good.
+int verify_checksum (char *ptr) {
+	int i,j,k;
+
+	for (i=j=0;((ptr[i]!=0) && (ptr[i]!='*'));++i) j^=ptr[i];
+	k=j&0xf;
+	if (ptr[i]!='*') return 0;
+	if (k<10) {
+		if (ptr[i+2] != '0'+k) return 0;
+	} else { 
+		if (ptr[i+2] != 'A'+k-10) return 0;
+	}
+	j=j>>4;
+	if (j<10) {
+		if (ptr[i+1] != '0'+j) return 0;
+	} else {
+		if (ptr[i+1] != 'A'+j-10) return 0;
+	}
+	return i;
+}
+
+void checksum (char *ptr) {
+	int i,j,k;
+
+	for (i=1,j=0,k=strlen(ptr);i<k;++i) j^=ptr[i];
+	sprintf(ptr+k,"*%02X",j);
+}
+
+inline void tx_mesg (char *client_message, int xcsoar_sock) {
+
+	checksum(client_message);
+	strcat(client_message,"\r\n");
+	if (send(xcsoar_sock, client_message, strlen(client_message), 0) < 0) 
+		fprintf(stderr, "send failed\n");
+}
+
+void parse_NMEA_sensor(char* message, t_sensor_context* sensors, int xcsoar_sock)
 { 
-	char *val;
-	char buffer[2001];
-	char delimiter[]=",*";
-	char *ptr=NULL;
-	
+	static char buffer[2001];
+	char del1[2]="$",del2[2]=",";
+	char *ptr=NULL,*ptr2,*last;
+	char tick=1;
+	char mesg[120];
+	int i;
+	static unsigned char counter=255;
+
 	// copy string and initialize strtok function
-	strncpy(buffer, message, strlen(message));
-	ptr = strtok(buffer, delimiter);	
-	
-	while (ptr != NULL)
-	{
-		
-		switch (*ptr)
-		{
-			case '$':
-			// skip start of NMEA sentence
-			break;
-			
-			case 'E':
-			// TE vario value
-			// get next value
-			ptr = strtok(NULL, delimiter);
-			val = (char *) malloc(strlen(ptr));
-			strncpy(val,ptr,strlen(ptr));
-			sensors->e = atof(val);
-			break;
-			
-			case 'Q':
-			// Airspeed value
-			// get next value
-			ptr = strtok(NULL, delimiter);
-			val = (char *) malloc(strlen(ptr));
-			strncpy(val,ptr,strlen(ptr));
-			sensors->q = atof(val);
-			break;
-			default:
-			break;
-		}
-		// get next part of string
-		ptr = strtok(NULL, delimiter);
+	if (status) {
+		if (strchr(buffer,'$')==NULL) counter++;
+		if (counter==0) {
+			strcpy(mesg,"$POV,?");
+			if (status&0x01) strcat(mesg,",POL");
```

Comment:
POL is not in the published protocol.
https://www.openvario.org/doku.php?id=projects:series_00:software:nmea

Reply:
I can remove it, I put it in because the $POV,C,POL was still in the code even though it was listed as deprecated.  I'm generally uncomfortable removing things because I worry someone will use it with an old version of XCSoar.

Reply:
Here you are **requesting** an answer.

Reply:
Only if you deliberately change the configuration to poll for it will it request it.  Currently it is set to only request the "Real Polar" and MacCready.  I can remove this if it really bothers you.